lib/evmone/eof.cpp

```diff
@@ -759,14 +765,17 @@ std::variant<EOF1Header, EOFValidationError> validate_header(
 
     const auto header_size = eof_header_size(section_headers);
 
-    const auto types_or_error =
-        validate_types(container, header_size, section_headers[TYPE_SECTION].front());
-    if (const auto* error = std::get_if<EOFValidationError>(&types_or_error))
-        return *error;
-    const auto& types = std::get<std::vector<EOFCodeType>>(types_or_error);
+    const auto type_section_offset = header_size;
+    const auto type_section_size = section_headers[TYPE_SECTION].front();
+
+    if (type_section_size != code_sizes.size() * 4)
+        return EOFValidationError::invalid_type_section_size;
+
+    auto validation_error = validate_types(container, type_section_offset, type_section_size);
```

Comment:
This is quite an issue: `validate_type()` should not be used in `validate_header()` and invoke separately after `validate_header()`. This way we can just pass `EOF1Header` to `validate_types()` and use its `.get_type()` method.

But maybe this requires a separate PR. What do you think @gumb0?